src/content/blog/content-layer-deep-dive.mdx

```diff
@@ -0,0 +1,230 @@
+
+A **collection** is defined as a set of entries that share a common schema. Each entry has a unique ID. It is analogous to a table in a relational database. When using the Astro Content Layer API, each collection uses a **loader**, which defines how the entries are loaded to populate that collection. A loader can be a basic inline function that returns an array of entries, or it can be a more advanced object that handles its own caching and data store, and may be distributed as a module on NPM. When a site is built, the loader for each collection is invoked, which will then update the local **data store**. The Astro pages can then query that data store using the `getCollection` and `getEntry` functions. This can be done during pre-rendering, or during server rendering if using an on-demand adapter. In each of these cases the same data is available, which is the snapshot from the time of the build.
+
+<BlogContentImage src={contentLayerArchitectureImage} />
+
+## Creating collections
+
+Content collections are defined in `src/content/config.ts`, and this is the same with the Content Layer API. The `loader` property defines the source of the data, with the simplest form of loader being an async function that returns an array of items:
+
+```ts
+import { defineCollection, z } from 'astro:content';
+
+const countries = defineCollection({
+	loader: async () => {
+		const response = await fetch('https://restcountries.com/v3.1/all');
+		const data = await response.json();
+		// Must return an array of entries with an id property, or an object with IDs as keys and entries as values
+		return data.map((country) => ({
+			id: country.cca3,
+			...country,
+		}));
+	},
+	// optionally define a schema using Zod
+	schema: z.object({
+		id: z.string(),
+		name: z.string(),
+		capital: z.array(z.string()),
+		population: z.number(),
+		// ...
+	}),
+});
+
+export const collections = { countries };
+```
+
+This data is then available in your `.astro` components.
+
+```astro
+---
+// src/pages/countries/[id].astro
+import type { GetStaticPaths } from 'astro';
+import { getCollection } from 'astro:content';
+
+export const getStaticPaths: GetStaticPaths = async () => {
+	const collection = await getCollection('countries');
+	if (!collection) return [];
+	return collection.map((country) => ({
+		params: {
+			id: country.id,
+		},
+		props: {
+			country,
+		},
+	}));
+};
+
+const { country } = Astro.props;
+---
+
+<h1>{craft.data.name}</h1>
+<p>Capital: {craft.data.capital}</p>
+```
+
```

Comment:
```suggestion

We were already pretty pleased with this experience of querying and rendering your data on the page, so we turned our attention to the mechanics underlying the process. Let's take that deep dive into how the Content Layer API organizes, manages, and uses your content!
```
A transition here could go nicely to recognize, "Yeah, it seems mostly the same so far when we explain how to use it, but here's what you came to read..."